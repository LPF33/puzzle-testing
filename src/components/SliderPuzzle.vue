<template>
    <h3>"{{ puzzleObj.title }}" puzzle (25 pieces)</h3>
    <h4>Elapsed Time: {{ currentElapsedTime }}</h4>
    <div v-if="checkForWin" class="user-win-game">
        <h3>You have solved the puzzle!!!</h3>
        <button @click="restart">Play again</button>
    </div>
    <div id="puzzle">
        <img
            v-for="(piece, index) of shuffledPuzzleArray"
            :key="index"
            :src="require(`../assets/${puzzleObj.id}/${piece}`)"
            @click="swapImg(index)"
        />
    </div>
    <button @click="closePuzzle" id="close-button">X</button>
</template>

<script>
const correctPuzzleArray = [
    "image_part_001.jpg",
    "image_part_002.jpg",
    "image_part_003.jpg",
    "image_part_004.jpg",
    "image_part_005.jpg",
    "image_part_006.jpg",
    "image_part_007.jpg",
    "image_part_008.jpg",
    "image_part_009.jpg",
    "image_part_010.jpg",
    "image_part_011.jpg",
    "image_part_012.jpg",
    "image_part_013.jpg",
    "image_part_014.jpg",
    "image_part_015.jpg",
    "image_part_016.jpg",
    "image_part_017.jpg",
    "image_part_018.jpg",
    "image_part_019.jpg",
    "image_part_020.jpg",
    "image_part_021.jpg",
    "image_part_022.jpg",
    "image_part_023.jpg",
    "image_part_024.jpg",
    "image_part_025.jpg",
];
export default {
    name: "SliderPuzzle",
    props: ["puzzleObj"],
    data() {
        return {
            correctPuzzleArray,
            shuffledPuzzleArray: [...correctPuzzleArray].sort(
                () => Math.random() - 0.5
            ),
            timerId: null,
            indexedToSwap: [],
            startDateTime: Date.now(),
            currentElapsedTime: null,
        };
    },
    mounted() {
        this.timerId = setInterval(this.elapsedTime, 1000);
    },
    beforeUnmount() {
        console.log("destroyed");
        clearInterval(this.timerId);
    },
    emits: ["close"],
    methods: {
        closePuzzle() {
            clearInterval(this.timerId);
            this.$emit("close");
        },
        elapsedTime() {
            const elpased = new Date(Date.now() - this.startDateTime);
            this.currentElapsedTime = elpased.toISOString().slice(14, 19);
        },
        swapImg(index) {
            if (this.indexedToSwap.length < 2) {
                this.indexedToSwap.push(index);
            }

            if (this.indexedToSwap.length === 2) {
                const [first, second] = this.indexedToSwap;
                [
                    this.shuffledPuzzleArray[first],
                    this.shuffledPuzzleArray[second],
                ] = [
                    this.shuffledPuzzleArray[second],
                    this.shuffledPuzzleArray[first],
                ];

                this.indexedToSwap = [];
            }
        },
        saveRecord() {
            try {
                const records =
                    JSON.parse(localStorage.getItem("records")) || [];
                records.push([
                    this.puzzleObj.id,
                    Date.now() - this.startDateTime,
                ]);
                localStorage.setItem("records", JSON.stringify(records));
            } catch (error) {
                console.log("could not save record");
            }
        },
        restart() {
            this.shuffledPuzzleArray = [...correctPuzzleArray].sort(
                () => Math.random() - 0.5
            );
            this.startDateTime = Date.now();
            this.timerId = setInterval(this.elapsedTime, 1000);
        },
    },
    computed: {
        checkForWin() {
            for (let i = 0; i < correctPuzzleArray.length; i++) {
                if (correctPuzzleArray[i] !== this.shuffledPuzzleArray[i]) {
                    return false;
                }
            }
            clearInterval(this.timerId);
            this.saveRecord();
            return true;
        },
    },
};
</script>
<style scoped>
#puzzle {
    display: grid;
    grid-template: repeat(5, minmax(50px, 100px)) / repeat(
            5,
            minmax(50px, 100px)
        );
    align-items: center;
    justify-content: center;
    margin-bottom: 10px;
}

#puzzle img {
    width: 100%;
    height: 100%;
}

#close-button {
    position: fixed;
    top: 10px;
    right: 10px;
    padding: 10px;
    border-radius: 5px;
    background-color: #766161;
    color: white;
    font-weight: bold;
}

h3 {
    color: #766161;
    text-shadow: 1px 1px 2px black;
    background-color: #e1f1dd;
    padding: 10px 0;
}

h4 {
    color: black;
    text-shadow: 1px 1px 2px black;
    padding: 10px 0;
}
</style>
