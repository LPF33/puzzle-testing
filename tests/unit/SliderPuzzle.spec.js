import { mount } from "@vue/test-utils";
import SliderPuzzle from "@/components/SliderPuzzle";
import "jest-localstorage-mock";
jest.useFakeTimers();

afterEach(() => {
    jest.clearAllMocks();
});

test("SliderPuzzle gets rendered with owl object correctly", () => {
    const wrapper = mount(SliderPuzzle, {
        props: {
            puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
        },
    });

    expect(wrapper.html()).toMatchSnapshot();
    afterEach(() => {
        jest.clearAllMocks();
    });
});

test("SliderPuzzle gets rendered with eggcozy object correctly", () => {
    const wrapper = mount(SliderPuzzle, {
        props: {
            puzzleObj: {
                id: "eggcozy",
                image: "eggcozy.jpg",
                title: "Egg cozy",
            },
        },
    });

    expect(wrapper.html()).toMatchSnapshot();
});

test("store index of image to swap when clicked on any image", () => {
    const wrapper = mount(SliderPuzzle, {
        props: {
            puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
        },
    });
    wrapper.find("img").trigger("click");
    expect(wrapper.vm.indexedToSwap.length).toBeGreaterThan(0);
});

const randomNum = () => {
    let nums = [];
    nums.push(Math.floor(Math.random() * 25));

    while (typeof nums[1] === "undefined" || nums[0] === nums[1]) {
        nums[1] = Math.floor(Math.random() * 25);
    }
    return nums;
};

test("swaps two images when 2 random images are clicked", async () => {
    const wrapper = mount(SliderPuzzle, {
        props: {
            puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
        },
    });
    const [firstNum, secondNum] = randomNum();
    const [firstImage, secondImage] = [
        wrapper.vm.shuffledPuzzleArray[firstNum],
        wrapper.vm.shuffledPuzzleArray[secondNum],
    ];
    wrapper.find(`#puzzle img:nth-child(${firstNum + 1})`).trigger("click");
    wrapper.find(`#puzzle img:nth-child(${secondNum + 1})`).trigger("click");
    expect(wrapper.vm.indexedToSwap.length).toBe(0);
    await wrapper.vm.$nextTick();
    const [newFirstImage, newSecondImage] = [
        wrapper.vm.shuffledPuzzleArray[firstNum],
        wrapper.vm.shuffledPuzzleArray[secondNum],
    ];
    expect(firstImage).toBe(newSecondImage);
    expect(secondImage).toBe(newFirstImage);
});

test("starts timer when Component mounted", () => {
    mount(SliderPuzzle, {
        props: {
            puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
        },
    });
    expect(setInterval).toHaveBeenCalled();
    expect(setInterval).toBeCalledTimes(1);
    expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
});

test("stops timer when close-button is clicked", () => {
    const wrapper = mount(SliderPuzzle, {
        props: {
            puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
        },
    });

    wrapper.find("#close-button").trigger("click");
    expect(clearInterval).toHaveBeenCalledTimes(1);
});

// test("stops timer when component is unmounted", async () => {
//     const wrapper = mount(SliderPuzzle, {
//         props: {
//             puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
//         },
//     });
//     await wrapper.vm.$destroy(); // .beforeUnmount does also not work
//     expect(clearInterval).toHaveBeenCalledTimes(1);
// });

test("save record to local storage", () => {
    const wrapper = mount(SliderPuzzle, {
        props: {
            puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
        },
        data() {
            return {
                startDateTime: new Date(),
                currentElapsedTime: new Date(),
            };
        },
    });
    wrapper.vm.saveRecord();
    const { startDateTime, puzzleObj } = wrapper.vm;
    const recordsToBeSaved = JSON.stringify([
        [puzzleObj.id, Date.now() - startDateTime],
    ]);
    expect(localStorage.setItem).toHaveBeenCalledWith(
        "records",
        recordsToBeSaved
    );
});

test("shows the elapsed time", () => {
    const wrapper = mount(SliderPuzzle, {
        props: {
            puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
        },
        data() {
            return {
                currentElapsedTime: "00:35",
            };
        },
    });
    expect(wrapper.html()).toContain("00:35");
});

test("user has won the game and can restart game", async () => {
    const wrapper = mount(SliderPuzzle, {
        props: {
            puzzleObj: { id: "owl", image: "owl.jpg", title: "Owl" },
        },
    });
    expect(wrapper.vm.checkForWin).toBeFalsy();
    await wrapper.vm.$nextTick();

    expect(clearInterval).toHaveBeenCalledTimes(0);
    wrapper.vm.shuffledPuzzleArray = [...wrapper.vm.correctPuzzleArray];
    expect(wrapper.vm.checkForWin).toBeTruthy();

    await wrapper.vm.$nextTick();
    expect(clearInterval).toHaveBeenCalledTimes(1);
    jest.clearAllMocks();
    expect(wrapper.find(".user-win-game").exists()).toBeTruthy();
    wrapper.find(".user-win-game button").trigger("click");
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(wrapper.html()).toMatchSnapshot();

    await wrapper.vm.$nextTick();
    expect(wrapper.find(".user-win-game").exists()).toBeFalsy();
    expect(wrapper.html()).toMatchSnapshot();
});
