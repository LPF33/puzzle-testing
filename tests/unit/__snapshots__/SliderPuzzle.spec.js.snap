// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderPuzzle gets rendered with eggcozy object correctly 1`] = `
<h3>"Egg cozy" puzzle (25 pieces)</h3>
<h4>Elapsed Time: </h4>
<!--v-if-->
<div id="puzzle"><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""></div><button id="close-button">X</button>
`;

exports[`SliderPuzzle gets rendered with owl object correctly 1`] = `
<h3>"Owl" puzzle (25 pieces)</h3>
<h4>Elapsed Time: </h4>
<!--v-if-->
<div id="puzzle"><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""></div><button id="close-button">X</button>
`;

exports[`user has won the game and can restart game 1`] = `
<h3>"Owl" puzzle (25 pieces)</h3>
<h4>Elapsed Time: </h4>
<div class="user-win-game">
  <h3>You have solved the puzzle!!!</h3><button>Play again</button>
</div>
<div id="puzzle"><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""></div><button id="close-button">X</button>
`;

exports[`user has won the game and can restart game 2`] = `
<h3>"Owl" puzzle (25 pieces)</h3>
<h4>Elapsed Time: </h4>
<!--v-if-->
<div id="puzzle"><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""><img src=""></div><button id="close-button">X</button>
`;
